Kernel/Bus/PCI/Controller/HostController.cpp

```diff
@@ -128,10 +148,120 @@ UNMAP_AFTER_INIT void HostController::enumerate_attached_devices(Function<void(E
                 continue;
-            enumerate_bus(callback, m_domain.start_bus() + bus_as_function_number, false);
+            enumerate_bus(callback, post_bridge_callback, m_domain.start_bus() + bus_as_function_number, false);
             m_enumerated_buses.set(m_domain.start_bus() + bus_as_function_number, true);
         }
     }
 }
 
+void HostController::configure_attached_devices(FlatPtr& mmio_32bit_base, FlatPtr mmio_32bit_end, FlatPtr& mmio_64bit_base, FlatPtr mmio_64bit_end)
+{
+    // First, Assign PCI-to-PCI bridge bus numbering
+    u8 bus_id = 0;
+    enumerate_attached_devices([this, &bus_id](EnumerableDeviceIdentifier const& device_identifier) {
+        // called for each PCI device encountered (including bridges)
+        if (read8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::CLASS) != PCI::ClassID::Bridge)
+            return;
+        if (read8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SUBCLASS) != PCI::Bridge::SubclassID::PCI_TO_PCI)
+            return;
+        write8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SECONDARY_BUS, ++bus_id);
+        write8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SUBORDINATE_BUS, 0xFF); }, [this, &bus_id](EnumerableDeviceIdentifier const& device_identifier) {
+        // called after a bridge was recursively enumerated
+        write8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SUBORDINATE_BUS, bus_id); });
+
+    // Second, Assign BAR addresses
+    // TODO: We currently naively assign addresses bump-allocator style - Switch to a proper allocator if this is not good enough
+    enumerate_attached_devices([this, &mmio_32bit_base, mmio_32bit_end, &mmio_64bit_base, mmio_64bit_end](EnumerableDeviceIdentifier const& device_identifier) {
+        // device-generic handling
+        auto header_type = read8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::HEADER_TYPE);
+        auto const max_bar = (header_type == 0) ? RegisterOffset::BAR5 : RegisterOffset::BAR1;
+        for (u32 bar_offset = to_underlying(RegisterOffset::BAR0); bar_offset <= to_underlying(max_bar); bar_offset += 4) {
+            auto bar_value = read32_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), bar_offset);
+            auto bar_space_type = get_BAR_space_type(bar_value);
+            auto bar_prefetchable = (bar_value >> 3) & 1;
```

Comment:
Do we need to verify that the prefetchable "p" bit of the PCI MMIO range is set when writing the value of a prefetchable BAR?

Reply:
I don't think so, the specification defines that: `A Function is permitted to mark a range as prefetchable if there are no side effects on reads, the Function returns all bytes on reads regardless of the byte enables, and host bridges can merge processor writes into this range without causing errors.`
Which I read as meaning that it _can_ use prefetchable memory, not that it _must_.

Reply:
I don't think so, the specification defines that: `A Function is permitted to mark a range as prefetchable if there are no side effects on reads, the Function returns all bytes on reads regardless of the byte enables, and host bridges can merge processor writes into this range without causing errors.`
Which I read as meaning that it _can_ use prefetchable memory, not that it _must_.

Reply:
None of the `ranges` entries have this bit set either, so I think this should be fine.

Reply:
Ah no, the bit probably just isn't set in qemu.

But wouldn't it be bad if we would write the address of a prefetchable region into a BAR that is marked as non-prefetchable?
Because the device then expects that the region won't prefetched.
I am not sure if the "p" bit in `ranges` means that a range will be prefetched though, the PCI binding is not entirely clear about that.

Reply:
I see what you mean, the inverted case of a prefetchable region given to a non-prefetchable BAR does sound like an issue. Though, since the PCI-to-PCI bridge specification lets you only specify 64 bit addresses as prefetchable, and doesn't let you specify both a 32-bit prefetchable and 32bit non-prefetchable range, it sounds like implicitly 32-bit ranges should _not_ be prefetchable.
So I guess I can add a check that we will only use non-prefetchable regions for the 32-bit range, and we can look into this further once we trip up on a device which doesn't have such a range.

Reply:
Yeah, I think this is fine then.
My VisionFive 2 also only has a prefetchable 64-bit region. The 32-bit region is non-prefetchable.